File: app/src/main/java/com/example/myapplication/core/catalog/CatalogValidator.kt
package com.example.myapplication.core.catalog

import com.example.myapplication.core.model.ExerciseDefinition
import com.example.myapplication.core.model.ProgramTemplate

object CatalogValidator {
    private val validId = Regex("[a-z0-9_]+")

    fun validateExercises(exercises: List<ExerciseDefinition>): List<String> {
        val issues = mutableListOf<String>()
        val exercisesById = exercises.associateBy(ExerciseDefinition::id)

        exercises.groupingBy { it.id }
            .eachCount()
            .filterValues { it > 1 }
            .keys
            .sorted()
            .forEach { issues += "Duplicate exercise id: $it" }

        exercises.forEach { exercise ->
            if (!validId.matches(exercise.id)) {
                issues += "Exercise id '${exercise.id}' must match [a-z0-9_]+"
            }
            if (exercise.sourceId.isBlank()) {
                issues += "Exercise '${exercise.id}' has blank sourceId"
            }
            if (exercise.nameVi.isBlank()) {
                issues += "Exercise '${exercise.id}' has blank nameVi"
            }
            if (exercise.instructionsVi.size !in 2..5) {
                issues += "Exercise '${exercise.id}' instructionsVi must contain 2..5 items"
            }
            if (exercise.instructionsVi.any { it.isBlank() }) {
                issues += "Exercise '${exercise.id}' has a blank instruction"
            }
            if (exercise.equipment.isEmpty()) {
                issues += "Exercise '${exercise.id}' must declare equipment"
            }
            exercise.substituteIds.groupingBy { it }.eachCount()
                .filterValues { it > 1 }
                .keys
                .sorted()
                .forEach { duplicate ->
                    issues += "Exercise '${exercise.id}' has duplicate substitute '$duplicate'"
                }
            exercise.substituteIds.distinct().forEach { substituteId ->
                when {
                    substituteId == exercise.id -> {
                        issues += "Exercise '${exercise.id}' cannot substitute itself"
                    }
                    substituteId !in exercisesById -> {
                        issues += "Exercise '${exercise.id}': Unknown substitute '$substituteId'"
                    }
                    exercisesById.getValue(substituteId).primaryMuscle != exercise.primaryMuscle -> {
                        issues += "Exercise '${exercise.id}' substitute '$substituteId' must use the same primary muscle"
                    }
                    exercisesById.getValue(substituteId).movementPattern != exercise.movementPattern -> {
                        issues += "Exercise '${exercise.id}' substitute '$substituteId' must use the same movement pattern"
                    }
                }
            }
        }

        return issues
    }

    fun validatePrograms(
        programs: List<ProgramTemplate>,
        exercisesById: Map<String, ExerciseDefinition>,
    ): List<String> {
        val issues = mutableListOf<String>()
        programs.groupingBy { it.id }.eachCount().filterValues { it > 1 }.keys.sorted().forEach {
            issues += "Duplicate program id: $it"
        }
        programs.groupingBy {
            listOf(it.goal, it.level, it.equipmentProfile, it.sessionsPerWeek, it.durationWeeks)
        }.eachCount().filterValues { it > 1 }.keys.forEach {
            issues += "Duplicate program match key: ${it.joinToString()}"
        }
        programs.forEach { program ->
            if (program.id.isBlank()) issues += "Program has blank id"
            if (program.sessionsPerWeek !in 1..7) {
                issues += "Program '${program.id}' sessionsPerWeek must be in 1..7"
            }
            if (program.durationWeeks !in 1..52) {
                issues += "Program '${program.id}' durationWeeks must be in 1..52"
            }
            val expectedWorkoutCount = program.sessionsPerWeek * program.durationWeeks
            if (program.workouts.size != expectedWorkoutCount) issues += "Program '${program.id}' workouts.size must be $expectedWorkoutCount"
            if (program.workouts.map { it.sequence }.sorted() != (0 until program.workouts.size).toList()) {
                issues += "Program '${program.id}' sequences must be contiguous starting at 0"
            }
            program.workouts.forEach { workout ->
                val workoutLabel = "Program '${program.id}' workout ${workout.sequence}"
                if (workout.titleVi.isBlank()) issues += "$workoutLabel has blank titleVi"
                if (workout.focusVi.isBlank()) issues += "$workoutLabel has blank focusVi"
                if (workout.exercises.isEmpty()) issues += "$workoutLabel must contain exercises"
                if (workout.week !in 1..program.durationWeeks) issues += "$workoutLabel week must be in 1..${program.durationWeeks}"
                if (workout.estimatedMinutes !in 10..90) issues += "$workoutLabel estimatedMinutes must be in 10..90"
                if (workout.restDaysAfter !in 0..3) issues += "$workoutLabel restDaysAfter must be in 0..3"
                workout.exercises.forEachIndexed { index, prescription ->
                    val exerciseLabel = "$workoutLabel exercise $index"
                    if (prescription.exerciseId !in exercisesById) issues += "$exerciseLabel: Unknown exercise '${prescription.exerciseId}'"
                    if (prescription.sets !in 1..6) issues += "$exerciseLabel sets must be in 1..6"
                    if (prescription.restSeconds !in 15..300) issues += "$exerciseLabel restSeconds must be in 15..300"
                    val validReps = prescription.repsMin != null && prescription.repsMax != null &&
                        prescription.repsMin in 1..50 && prescription.repsMax in prescription.repsMin..100 &&
                        prescription.durationSeconds == null
                    val validDuration = prescription.durationSeconds != null && prescription.durationSeconds in 10..3600 &&
                        prescription.repsMin == null && prescription.repsMax == null
                    if (validReps == validDuration) issues += "$exerciseLabel prescription must use exactly one valid reps or duration mode"
                }
            }
            (1..program.durationWeeks).forEach { week ->
                val weeklyWorkouts = program.workouts.filter { it.week == week }
                if (weeklyWorkouts.size != program.sessionsPerWeek) issues += "Program '${program.id}' week $week must contain ${program.sessionsPerWeek} workouts"
                if (weeklyWorkouts.size + weeklyWorkouts.sumOf { it.restDaysAfter } != 7) {
                    issues += "Program '${program.id}' week $week weekly schedule must total 7 days"
                }
            }
        }
        return issues
    }
}

File: flutter/lib/core/catalog/catalog_validator.dart
import '../model/catalog_models.dart';
import '../model/movement_block_models.dart';

class CatalogValidator {
  static final _validId = RegExp(r'^[a-z0-9_]+$');

  static List<String> validateMovementBlocks(List<MovementBlock> blocks) {
    final issues = <String>[];

    final idCounts = <String, int>{};
    for (var b in blocks) {
      idCounts[b.id] = (idCounts[b.id] ?? 0) + 1;
    }
    final duplicateIds =
        idCounts.entries.where((e) => e.value > 1).map((e) => e.key).toList()
          ..sort();

    for (var id in duplicateIds) {
      issues.add("Duplicate movement block id: $id");
    }

    for (var block in blocks) {
      if (!_validId.hasMatch(block.id)) {
        issues.add("Movement block id '${block.id}' must match [a-z0-9_]+");
      }
      if (block.titleVi.trim().isEmpty) {
        issues.add("Movement block '${block.id}' has blank titleVi");
      }
      if (block.movementPatterns.isEmpty) {
        issues.add(
          "Movement block '${block.id}' movementPatterns must not be empty",
        );
      }
      if (block.stepsVi.length < 2 || block.stepsVi.length > 6) {
        issues.add(
          "Movement block '${block.id}' stepsVi must contain 2..6 items",
        );
      }
      if (block.stepsVi.any((s) => s.trim().isEmpty)) {
        issues.add("Movement block '${block.id}' has a blank step");
      }
      if (block.estimatedMinutes < 2 || block.estimatedMinutes > 10) {
        issues.add(
          "Movement block '${block.id}' estimatedMinutes must be in 2..10",
        );
      }
    }
    return issues;
  }

  static List<String> validateExercises(List<ExerciseDefinition> exercises) {
    final issues = <String>[];
    final exercisesById = {for (var e in exercises) e.id: e};

    final idCounts = <String, int>{};
    for (var e in exercises) {
      idCounts[e.id] = (idCounts[e.id] ?? 0) + 1;
    }
    final duplicateIds =
        idCounts.entries.where((e) => e.value > 1).map((e) => e.key).toList()
          ..sort();

    for (var id in duplicateIds) {
      issues.add("Duplicate exercise id: $id");
    }

    for (var exercise in exercises) {
      if (!_validId.hasMatch(exercise.id)) {
        issues.add("Exercise id '${exercise.id}' must match [a-z0-9_]+");
      }
      final animationId = exercise.animationId;
      if (animationId != null && !_validId.hasMatch(animationId)) {
        issues.add(
          "Exercise '${exercise.id}' animationId must match [a-z0-9_]+",
        );
      }
      if (exercise.sourceId.trim().isEmpty) {
        issues.add("Exercise '${exercise.id}' has blank sourceId");
      }
      if (exercise.nameVi.trim().isEmpty) {
        issues.add("Exercise '${exercise.id}' has blank nameVi");
      }
      if (exercise.instructionsVi.length < 2 ||
          exercise.instructionsVi.length > 5) {
        issues.add(
          "Exercise '${exercise.id}' instructionsVi must contain 2..5 items",
        );
      }
      if (exercise.instructionsVi.any((s) => s.trim().isEmpty)) {
        issues.add("Exercise '${exercise.id}' has a blank instruction");
      }
      if (exercise.equipment.isEmpty) {
        issues.add("Exercise '${exercise.id}' must declare equipment");
      }

      final subCounts = <String, int>{};
      for (var sub in exercise.substituteIds) {
        subCounts[sub] = (subCounts[sub] ?? 0) + 1;
      }
      final duplicateSubs =
          subCounts.entries.where((e) => e.value > 1).map((e) => e.key).toList()
            ..sort();

      for (var duplicate in duplicateSubs) {
        issues.add(
          "Exercise '${exercise.id}' has duplicate substitute '$duplicate'",
        );
      }

      final uniqueSubs = exercise.substituteIds.toSet();
      for (var substituteId in uniqueSubs) {
        if (substituteId == exercise.id) {
          issues.add("Exercise '${exercise.id}' cannot substitute itself");
        } else if (!exercisesById.containsKey(substituteId)) {
          issues.add(
            "Exercise '${exercise.id}': Unknown substitute '$substituteId'",
          );
        } else {
          final subDefinition = exercisesById[substituteId]!;
          if (subDefinition.primaryMuscleGroup != exercise.primaryMuscleGroup) {
            issues.add(
              "Exercise '${exercise.id}' substitute '$substituteId' must use the same primary muscle",
            );
          }
          if (subDefinition.movementPattern != exercise.movementPattern) {
            issues.add(
              "Exercise '${exercise.id}' substitute '$substituteId' must use the same movement pattern",
            );
          }
        }
      }
    }

    return issues;
  }

  static List<String> validatePrograms(
    List<ProgramTemplate> programs,
    Map<String, ExerciseDefinition> exercisesById,
  ) {
    final issues = <String>[];

    final idCounts = <String, int>{};
    for (var p in programs) {
      idCounts[p.id] = (idCounts[p.id] ?? 0) + 1;
    }
    final duplicateIds =
        idCounts.entries.where((e) => e.value > 1).map((e) => e.key).toList()
          ..sort();

    for (var id in duplicateIds) {
      issues.add("Duplicate program id: $id");
    }

    final keyCounts = <String, int>{};
    for (var p in programs) {
      final key =
          "${p.goal.name}_${p.level.name}_${p.equipmentProfile.name}_${p.sessionsPerWeek}_${p.durationWeeks}";
      keyCounts[key] = (keyCounts[key] ?? 0) + 1;
    }
    final duplicateKeys = keyCounts.entries
        .where((e) => e.value > 1)
        .map((e) => e.key)
        .toList();
    for (var key in duplicateKeys) {
      issues.add("Duplicate program match key: ${key.split('_').join(', ')}");
    }

    for (var program in programs) {
      if (program.id.trim().isEmpty) {
        issues.add("Program has blank id");
      }
      if (program.sessionsPerWeek < 1 || program.sessionsPerWeek > 7) {
        issues.add("Program '${program.id}' sessionsPerWeek must be in 1..7");
      }
      if (program.durationWeeks < 1 || program.durationWeeks > 52) {
        issues.add("Program '${program.id}' durationWeeks must be in 1..52");
      }

      final expectedWorkoutCount =
          program.sessionsPerWeek * program.durationWeeks;
      if (program.workouts.length != expectedWorkoutCount) {
        issues.add(
          "Program '${program.id}' workouts.size must be $expectedWorkoutCount",
        );
      }

      final sequences = program.workouts.map((w) => w.sequence).toList()
        ..sort();
      final expectedSequences = List.generate(
        program.workouts.length,
        (i) => i,
      );

      bool contiguous = true;
      if (sequences.length != expectedSequences.length) {
        contiguous = false;
      } else {
        for (var i = 0; i < sequences.length; i++) {
          if (sequences[i] != expectedSequences[i]) {
            contiguous = false;
            break;
          }
        }
      }
      if (!contiguous) {
        issues.add(
          "Program '${program.id}' sequences must be contiguous starting at 0",
        );
      }

      for (var workout in program.workouts) {
        final workoutLabel =
            "Program '${program.id}' workout ${workout.sequence}";
        if (workout.titleVi.trim().isEmpty) {
          issues.add("$workoutLabel has blank titleVi");
        }
        if (workout.focusVi.trim().isEmpty) {
          issues.add("$workoutLabel has blank focusVi");
        }
        if (workout.exercises.isEmpty) {
          issues.add("$workoutLabel must contain exercises");
        }
        if (workout.week < 1 || workout.week > program.durationWeeks) {
          issues.add(
            "$workoutLabel week must be in 1..${program.durationWeeks}",
          );
        }
        if (workout.estimatedMinutes < 10 || workout.estimatedMinutes > 90) {
          issues.add("$workoutLabel estimatedMinutes must be in 10..90");
        }
        if (workout.restDaysAfter < 0 || workout.restDaysAfter > 3) {
          issues.add("$workoutLabel restDaysAfter must be in 0..3");
        }

        for (var i = 0; i < workout.exercises.length; i++) {
          final prescription = workout.exercises[i];
          final exerciseLabel = "$workoutLabel exercise $i";

          if (!exercisesById.containsKey(prescription.exerciseId)) {
            issues.add(
              "$exerciseLabel: Unknown exercise '${prescription.exerciseId}'",
            );
          }
          if (prescription.sets < 1 || prescription.sets > 6) {
            issues.add("$exerciseLabel sets must be in 1..6");
          }
          if (prescription.restSeconds < 15 || prescription.restSeconds > 300) {
            issues.add("$exerciseLabel restSeconds must be in 15..300");
          }

          final validReps =
              prescription.minReps != null &&
              prescription.maxReps != null &&
              prescription.minReps! >= 1 &&
              prescription.minReps! <= 50 &&
              prescription.maxReps! >= prescription.minReps! &&
              prescription.maxReps! <= 100 &&
              prescription.durationSeconds == null;

          final validDuration =
              prescription.durationSeconds != null &&
              prescription.durationSeconds! >= 10 &&
              prescription.durationSeconds! <= 3600 &&
              prescription.minReps == null &&
              prescription.maxReps == null;

          if (validReps == validDuration) {
            issues.add(
              "$exerciseLabel prescription must use exactly one valid reps or duration mode",
            );
          }
        }
      }

      for (var week = 1; week <= program.durationWeeks; week++) {
        final weeklyWorkouts = program.workouts
            .where((w) => w.week == week)
            .toList();
        if (weeklyWorkouts.length != program.sessionsPerWeek) {
          issues.add(
            "Program '${program.id}' week $week must contain ${program.sessionsPerWeek} workouts",
          );
        }

        final totalDays =
            weeklyWorkouts.length +
            weeklyWorkouts.fold<int>(0, (sum, w) => sum + w.restDaysAfter);
        if (totalDays != 7) {
          issues.add(
            "Program '${program.id}' week $week weekly schedule must total 7 days",
          );
        }
      }
    }
    return issues;
  }
}
